### ISSUE TITLE
Make release builds the default

### ISSUE THREAD

**Issue by pgkeller** [tenstorrent-metal/tt-metal#4255]
Presently, the default build is an assert build.  The default should be the peak performance build, release, in particular for CI perf testing.  Correctness tests could run w/ assert.

I recently got release builds working, but they regressed as developers use TT_ASSERT in the "default" statement of a switch.  Some of those cases fail to compile without the TT_ASSERT due to errors such as "variable may be used uninitialized in this fn".

Flipping to release as default will prevent these regressions from showing up again.

fyi: @tt-rkim

**Comment by TT-billteng** [comment]
I believe there were runtime crashes that need to be resolved

**Comment by tt-rkim** [comment]
CI builds are the default in CI, which is release build with `DEBUG` symbols.

We could try changing to `release` build for perf pipelines and see what kind of gains we get there>?

**Comment by TT-billteng** [comment]
Here's a test run with release build https://github.com/tenstorrent/tt-metal/actions/runs/8808018734/job/24176285444

**Comment by pgkeller** [comment]
Looks pretty bad.  Can we try a run w/o -lto to see if it is lto or other?  Not clear to me why there would be so many issues, I even saw one mentioning an "assert" failure which...should be gone?

**Comment by vtangTT** [comment]
Something I just noticed while working with CMake: building with LTO enabled, or -flto flag set will break some tests 

I accidentally built ttnn pybinds with -flto and `test_print_tensor`, test_reports` broke with assert failures.

**Comment by vtangTT** [comment]
here is a passing post commit with release build: https://github.com/tenstorrent/tt-metal/actions/runs/8950170132

Failures were bc: 
- `TT_ASSERT` is not defined for release build, so every test relying on that failed
- a validate step in ttnn's `run_operation` is not defined, so tests such as `tests/ttnn/unit_tests/test_validate_decorator.py` would fail

Not sure if we want to define these in release build or keep them as `DEBUG` only and find another way?

**Comment by pgkeller** [comment]
How do tests relay on TT_ASSERT?  That's broken.  Asserts go away in release builds.  If they need to flag an error, they need to use TT_FATAL.

**Comment by vtangTT** [comment]
An example would be `tests/tt_metal/tt_metal/unit_tests/host_apis/test_tilize_untilize.cpp` tests such as `TestTilizeThrowErrorForInvalidTileMandN` that relies on the `TT_ASSERT` inside `tt_metal/common/tilize_untilize.hpp` to throw an exception to get caught by a `EXPECT_ANY_THROW`

```TT_ASSERT(input.size() > 0 and m > 0 and n > 0, "None of the input size, m, nor n can be 0");```
```TT_ASSERT((input.size() % (m * n)) == 0, "Input size must be divisible by m  and n");```

**Comment by pgkeller** [comment]
Validating API inputs must be done w/ TT_FATAL, not assert.  Asserts are for internally generated values when you want to confirm some invariant.  Looks like the code in tileize_untilize should be changed.

**Comment by tt-rkim** [comment]
I think we need an effort to go and scan runtime code to change any production input validation from `TT_ASSERT` to `TT_FATAL` per @pgkeller 
Can we bump this up on priority to enable release builds (ie. more perf!!) ? 

cc: @TT-billteng

**Comment by pgkeller** [comment]
fwiw, I did this once a few months back, I suspect it isn't too far off what we need, however, we aren't consistent on how we handle this which makes the task harder...

**Comment by vtangTT** [comment]
should be done in this PR: https://github.com/tenstorrent/tt-metal/pull/8176

**Comment by vtangTT** [comment]
device perf in `Release`: https://github.com/tenstorrent/tt-metal/actions/runs/8987453667
models perf in `Release`: https://github.com/tenstorrent/tt-metal/actions/runs/8987456170

There was a slight improvement in models perf on `Release` build.

**Comment by davorchap** [comment]
> device perf in `Release`: https://github.com/tenstorrent/tt-metal/actions/runs/8987453667
> 
> models perf in `Release`: https://github.com/tenstorrent/tt-metal/actions/runs/8987456170
> 
> 
> 
> There was a slight improvement in models perf on `Release` build.

We should update the targets

This doesn't include the performance governor yet ?

**Comment by tt-rkim** [comment]
No, performance governor is a machine dependent thing and @TT-billteng and @tapspatel are working on that.

Orthogonal to @vtangTT 's work here

**Comment by vtangTT** [comment]
Release build is fixed in main: https://github.com/tenstorrent/tt-metal/pull/8176
